monitoring/storage.go

```
@@ -0,0 +1,238 @@
+	// WillBeCreated when true, then all checks will be applied to first existing dir, or fail otherwise
+	WillBeCreated bool
+	// path is normalized path, representing parent directory
+	// in case Path does not exist yet
+	path string
+	// MinBytesPerSecond is minimum write speed for probe to succeed
+	MinBytesPerSecond uint64
+	// Filesystems define list of supported filesystems, or any if empty
+	Filesystems []string
+	// MinFreeBytes define minimum free volume capacity
+	MinFreeBytes uint64
+}
+
+const (
+	storageWriteCheckerID = "io-check"
+	blockSize             = 1e5
+	cycles                = 1024
+	stRdonly              = int64(1)
+)
+
+// Name returns name of the checker
+func (c *StorageChecker) Name() string {
+	return fmt.Sprintf("%s(%s)", storageWriteCheckerID, c.Path)
+}
+
+func (c *StorageChecker) Check(ctx context.Context, reporter health.Reporter) {
+	err := c.check(ctx, reporter)
+	if err != nil {
+		reporter.Add(NewProbeFromErr(c.Name(), "internal error", trace.Wrap(err)))
+	} else {
+		reporter.Add(&pb.Probe{Checker: c.Name(), Status: pb.Probe_Running})
+	}
+}
+
+func (c *StorageChecker) check(ctx context.Context, reporter health.Reporter) error {
+	err := c.evalPath()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return trace.NewAggregate(c.checkFsType(ctx, reporter),
+		c.checkCapacity(ctx, reporter),
+		c.checkWriteSpeed(ctx, reporter))
+}
+
+// cleanPath returns fully evaluated path with symlinks resolved
+// if path doesn't exist but should be created, then it returns
+// first available parent directory, and checks will be applied to it
+func (c *StorageChecker) evalPath() error {
+	fi, err := os.Stat(c.Path)
+
+	if os.IsNotExist(err) && !c.WillBeCreated {
+		return trace.BadParameter("%s does not exist", c.Path)
+	}
+
+	p := c.Path
+	for {
+		fi, err = os.Stat(p)
+		if err == nil && fi.IsDir() {
+			c.path = p
+			return nil
+		} else if err == nil && !fi.IsDir() {
+			return trace.BadParameter("%s is not a directory", p)
+		}
```

Non-nil err is ignored.

---

yes, deliberately; we just go one level up.